p2p/src/error.rs

```diff
@@ -14,124 +14,133 @@
-    PublishError(String),
-    #[error("IdentifyError: `{0:?}`")]
-    IdentifyError(String),
+pub enum PeerError {
+    #[error("Peer disconnected")]
+    PeerDisconnected,
+    #[error("No peers")]
+    NoPeers,
+    #[error("Peer doesn't exist")]
+    PeerDoesntExist,
+    #[error("Peer already exists")]
+    PeerExists,
+}
+
+#[derive(Error, Debug, PartialEq, Eq)]
+pub enum PublishError {
+    #[error("Message has already been published")]
+    Duplicate,
+    #[error("Failed to sign message")]
+    SigningFailed,
+    #[error("Not enough peers in topic")]
+    InsufficientPeers,
+    #[error("Message is too large")]
+    MessageTooLarge,
+    #[error("Failed to compress the message")]
+    TransformFailed,
+}
+
+#[derive(Error, Debug, PartialEq, Eq)]
+pub enum DialError {
+    #[error("Peer is banned")]
+    Banned,
+    #[error("Limit for outgoing connections reached")]
+    ConnectionLimit,
+    #[error("Tried to dial local node")]
+    LocalPeerId,
+    #[error("Peer doesn't have any known addresses")]
+    NoAddresses,
+    #[error("Peer state not correct for dialing")]
+    DialPeerConditionFalse,
+    #[error("Connection has been aborted")]
+    Aborted,
+    #[error("Invalid PeerId")]
+    InvalidPeerId,
+    #[error("PeerId doesn't match the PeerId of endpoint")]
+    WrongPeerId,
+    #[error("I/O error: `{0:?}`")]
+    IoError(std::io::ErrorKind),
+    #[error("Failed to negotiate transport protocol")]
+    Transport,
+}
+
+#[derive(Error, Debug, PartialEq, Eq)]
+pub enum ConnectionError {
+    #[error("Timeout")]
+    Timeout,
+    #[error("Timer failed")]
+    Timer,
+    #[error("Failed to upgrade protocol")]
+    Upgrade,
 }
 
 #[derive(Error, Debug, PartialEq, Eq)]
 pub enum P2pError {
-    #[error("SocketError: `{0:?}`")]
-    SocketError(std::io::ErrorKind),
-    #[error("PeerDisconnected")]
-    PeerDisconnected,
-    #[error("DecodeFailure: `{0:?}`")]
-    DecodeFailure(String),
-    #[error("ProtocolError: `{0:?}`")]
+    #[error("Protocol violation: `{0:?}`")]
     ProtocolError(ProtocolError),
-    #[error("TimeError: `{0:?}`")]
-    TimeError(String),
-    #[error("Libp2pError: `{0:?}`")]
-    Libp2pError(Libp2pError),
-    #[error("Unknown: `{0:?}`")]
-    Unknown(String),
-    #[error("ChannelClosed")]
+    #[error("Failed to publish message: `{0:?}`")]
+    PublishError(PublishError),
+    #[error("Failed to upgrade connection: `{0:?}`")]
+    ConnectionError(ConnectionError),
+    #[error("Failed to dial peer: `{0:?}`")]
+    DialError(DialError),
+    #[error("Connection to other task lost")]
     ChannelClosed,
-    #[error("NoPeers")]
-    NoPeers,
-    #[error("PeerDoesntExist")]
-    PeerDoesntExist,
-    #[error("InvalidAddress")]
-    InvalidAddress,
-    #[error("InvalidData")]
-    InvalidData,
-    #[error("PeerExists")]
-    PeerExists,
+    #[error("Peer-related error: `{0:?}`")]
+    PeerError(PeerError),
+    #[error("Invalid data: `{0}`")]
+    InvalidData(&'static str),
```

Comment:
I actually hate seeing that `to_string()` so much that I just added a new subcategory for conversion errors so that I don't have to use it. `P2pError::Other` is the only category that uses `&'static str` and it's reserved for errors that are fatal but don't belong to any other category and don't have any meaningful parameters to add to the error message. It's like the `DispatchError::Other` from Substrate.

`InvalidData` is bad error because you could argue that almost all errors are `InvalidData` as long as you define `Invalid` properly. It's gone.